halo2_proofs/src/plonk/evaluation.rs

```diff
@@ -492,33 +508,127 @@ impl<C: CurveAffine> Evaluator<C> {
                         });
                     }
 
+                    // For lookups, compute inputs_inv_sum = ∑ 1 / (f_i(X) + α)
+                    // The outer vector has capacity self.lookups.len()
+                    // The middle vector has capacity domain.extended_len()
+                    // The inner vector has capacity 
+                    let inputs_inv_sum: Vec<Vec<Vec<_>>> = lookups
+                        .iter()
+                        .enumerate()
+                        .map(|(n, _)| {
+                            let (inputs_lookup_evaluator, _) = &self.lookups[n];
+                            let mut inputs_eval_data: Vec<_> = inputs_lookup_evaluator
+                                .iter()
+                                .map(|input_lookup_evaluator| input_lookup_evaluator.instance())
+                                .collect();
+
+                            let mut inputs_values_for_extended_domain: Vec<C::Scalar> =
+                                Vec::with_capacity(self.lookups[n].0.len() << domain.k());
+                            for idx in 0..size {
+                                // For each compressed input column, evaluate at ω^i and add beta
+                                // This is a vector of length self.lookups[n].0.len()
+                                let inputs_values: Vec<C::ScalarExt> = inputs_lookup_evaluator
+                                    .iter()
+                                    .zip(inputs_eval_data.iter_mut())
+                                    .map(|(input_lookup_evaluator, input_eval_data)| {
+                                        input_lookup_evaluator.evaluate(
+                                            input_eval_data,
+                                            fixed,
+                                            advice,
+                                            instance,
+                                            challenges,
+                                            &beta,
+                                            &gamma,
+                                            &theta,
+                                            &y,
+                                            &C::ScalarExt::zero(),
+                                            idx,
+                                            rot_scale,
+                                            isize,
+                                        )
+                                    })
+                                    .collect();
+
+                                inputs_values_for_extended_domain.extend_from_slice(&inputs_values);
+                            }
+
+                            inputs_values_for_extended_domain.batch_invert();
```

Comment:
[perf-1]
[parallel batch invert](https://github.com/scroll-tech/halo2/commit/0b535ebe0f00d257eff3742b1550fb9a148c119d) saved 40*8=320s in phase4, better to have this optimization

Reply:
@Velaciela can you push the change here?